src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,190 @@
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class MakeFieldReadonlyDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public MakeFieldReadonlyDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.ProjectAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                bool IsCandidateField(IFieldSymbol symbol) =>
+                        symbol.DeclaredAccessibility == Accessibility.Private &&
+                        !symbol.IsReadOnly &&
+                        !symbol.IsConst &&
+                        !symbol.IsImplicitlyDeclared &&
+                        symbol.Locations.Length == 1 &&
+                        !IsMutableValueType(symbol.Type);
+
+                // Stores the state for fields:
+                //  'isCandidate' : Indicates whether the field is a candidate to be made readonly based on it's declaration and options.
+                //  'written'     : Indicates if there are any writes to the field outside the constructor and field initializer.
+                var fieldStateMap = new ConcurrentDictionary<IFieldSymbol, (bool isCandidate, bool written)>();
+
+                void OnFieldWrite(IFieldSymbol field)
+                {
+                    Debug.Assert(IsCandidateField(field));
+                    Debug.Assert(fieldStateMap.ContainsKey(field));
+
+                    fieldStateMap[field] = (isCandidate: true, written: true);
+                }
+
+                (bool isCandidate, bool written) TryGetOrComputeIsCandidateField(IFieldSymbol fieldSymbol, AnalyzerOptions options, CancellationToken cancellationToken)
+                {
+                    return fieldStateMap.GetOrAdd(fieldSymbol, valueFactory: ComputeIsCandidate);
+
+                    (bool isCandidate, bool written) ComputeIsCandidate(IFieldSymbol field)
```

Comment:
i'm not a huge fan of excessive nesting of functions.  it makes flow control harder at times.

Reply:
We are capturing couple of parameters from the enclosing local function.

Reply:
i realize that :)  i'm just saying that excessive nesting makes things less clear.  it can be nicer for readability to keep things simpler.  My personal preference is really to not go deeper than one level of local functions.